File: katas/typescript/src/CleanCode/format/format-bad.ts
// Violación de Clean Code: Mal Formato
// ❌ Problemas: Formato horizontal y vertical incorrecto

// ❌ Líneas demasiado largas (formato horizontal malo)
class ProductService {
  private products: any[] = [
    { id: 1, name: "Laptop", price: 1200, category: "Electronics", stock: 50, supplier: "TechCorp", warranty: "2 years" },
    { id: 2, name: "Mouse", price: 25, category: "Accessories", stock: 200, supplier: "TechCorp", warranty: "1 year" },
  ];
  public findProductByIdAndUpdateStockAndCalculateDiscountAndSendNotification(id: number, quantity: number, discountRate: number): any {
    const product = this.products.find((p) => p.id === id);
    if (product) {
      product.stock -= quantity;
      const discountedPrice = product.price * discountRate;
      console.log(`Producto ${product.name} actualizado. Stock: ${product.stock}, Precio con descuento: ${discountedPrice}`);
      return { ...product, discountedPrice };
    }
    return null;
  }
}

// ❌ Sin espaciado vertical - todo junto
class OrderProcessor {
  private orders: any[] = [];
  public processOrder(customerId: number, items: any[], paymentMethod: string): boolean {
    if (!customerId || !items || items.length === 0) {
      return false;
    }
    let total = 0;
    for (let i = 0; i < items.length; i++) {
      total += items[i].price * items[i].quantity;
    }
    if (total > 100) {
      total = total * 0.9;
    }
    const order = { id: Date.now(), customerId, items, total, date: new Date(), status: "pending" };
    this.orders.push(order);
    console.log(`Orden procesada: $${total}`);
    return true;
  }
  public getOrders(): any[] {
    return this.orders;
  }
  public cancelOrder(orderId: number): boolean {
    const index = this.orders.findIndex((o) => o.id === orderId);
    if (index >= 0) {
      this.orders.splice(index, 1);
      return true;
    }
    return false;
  }
}

// ❌ Formato inconsistente y mezcla de estilos
class UserManager {
  private users: any[] = [];
  public addUser(name: string, email: string, age: number) {
    const user = { name: name, email: email, age: age, createdAt: new Date() };
    this.users.push(user);
    return user;
  }

  public findUser(email: string) {
    for (let i = 0; i < this.users.length; i++) {
      if (this.users[i].email === email) {
        return this.users[i];
      }
    }
    return null;
  }

  public deleteUser(email: string) {
    this.users = this.users.filter((u: any) => u.email !== email);
  }
}

// Uso
console.log("=== Violación de Formato en Clean Code ===");
const service = new ProductService();
const product = service.findProductByIdAndUpdateStockAndCalculateDiscountAndSendNotification(1, 5, 0.8);
console.log(product);
const processor = new OrderProcessor();
processor.processOrder(1, [{ price: 100, quantity: 2 }], "credit_card");

export { ProductService, OrderProcessor, UserManager };

File: katas/python/CleanCode/format/format_bad.py
# Violación de Clean Code: Mal Formato
# ❌ Problemas: Formato horizontal y vertical incorrecto

from typing import Dict, List, Any
from datetime import datetime


# ❌ Líneas demasiado largas (formato horizontal malo)
class ProductService:
    def __init__(self):
        self.products: List[Dict[str, Any]] = [
            {"id": 1, "name": "Laptop", "price": 1200, "category": "Electronics",
                "stock": 50, "supplier": "TechCorp", "warranty": "2 years"},
            {"id": 2, "name": "Mouse", "price": 25, "category": "Accessories",
                "stock": 200, "supplier": "TechCorp", "warranty": "1 year"},
        ]

    def find_product_by_id_and_update_stock_and_calculate_discount_and_send_notification(
        self, product_id: int, quantity: int, discount_rate: float
    ) -> Dict[str, Any] | None:
        product = next(
            (p for p in self.products if p["id"] == product_id), None)
        if product:
            product["stock"] -= quantity
            discounted_price = product["price"] * discount_rate
            print(
                f"Producto {product['name']} actualizado. Stock: {product['stock']}, Precio con descuento: {discounted_price}")
            return {**product, "discounted_price": discounted_price}
        return None


# ❌ Sin espaciado vertical - todo junto
class OrderProcessor:
    def __init__(self):
        self.orders: List[Dict[str, Any]] = []

    def process_order(self, customer_id: int, items: List[Dict[str, Any]], payment_method: str) -> bool:
        if not customer_id or not items or len(items) == 0:
            return False
        total = 0
        for item in items:
            total += item["price"] * item["quantity"]
        if total > 100:
            total = total * 0.9
        order = {"id": datetime.now().timestamp(), "customer_id": customer_id,
                 "items": items, "total": total, "date": datetime.now(), "status": "pending"}
        self.orders.append(order)
        print(f"Orden procesada: ${total}")
        return True

    def get_orders(self) -> List[Dict[str, Any]]:
        return self.orders

    def cancel_order(self, order_id: float) -> bool:
        self.orders = [o for o in self.orders if o["id"] != order_id]
        return True


# ❌ Formato inconsistente y mezcla de estilos
class UserManager:
    def __init__(self):
        self.users: List[Dict[str, Any]] = []

    def add_user(self, name: str, email: str, age: int):
        user = {"name": name, "email": email,
                "age": age, "created_at": datetime.now()}
        self.users.append(user)
        return user

    def find_user(self, email: str):
        for user in self.users:
            if user["email"] == email:
                return user
        return None

    def delete_user(self, email: str):
        self.users = [u for u in self.users if u["email"] != email]


if __name__ == "__main__":
    # Uso
    print("=== Violación de Formato en Clean Code ===")
    service = ProductService()
    product = service.find_product_by_id_and_update_stock_and_calculate_discount_and_send_notification(
        1, 5, 0.8)
    print(product)
    processor = OrderProcessor()
    processor.process_order(1, [{"price": 100, "quantity": 2}], "credit_card")
